File: day-23/part-2/thomren.py
from collections import defaultdict, deque
from functools import cache
from tool.runners.python import SubmissionPy


class ThomrenSubmission(SubmissionPy):
    def run(self, s: str):
        """
        :param s: input in string format
        :return: solution flag
        """
        grid = [[c for c in line] for line in s.split("\n")]
        start = (0, next(i for i, c in enumerate(grid[0]) if c == "."))
        end = (len(grid) - 1, next(i for i, c in enumerate(grid[-1]) if c == "."))

        # build condensed graph
        stack = [start]
        visited = set()
        graph = defaultdict(set)
        while stack:
            pos = stack.pop()
            if pos == end or pos in visited:
                continue
            visited.add(pos)

            for neighbor in get_neighbors(grid, pos):
                (nx, ny), nd = find_next_intersection(
                    grid, neighbor, (neighbor[0] - pos[0], neighbor[1] - pos[1])
                )
                graph[pos].add(((nx, ny), nd))
                stack.append((nx, ny))

        # find longest path
        stack = [(start, 0)]
        visited = set()
        max_dist = 0
        while stack:
            pos, dist = stack.pop()
            if pos == end:
                max_dist = max(max_dist, dist)
                continue
            elif dist == -1:
                visited.remove(pos)
                continue
            elif pos in visited:
                continue
            visited.add(pos)

            # will backtrack once all the paths from pos have been explored
            stack.append((pos, -1))

            for neighbor, d in graph[pos]:
                if neighbor not in visited:
                    stack.append((neighbor, dist + d))

        return max_dist


def find_next_intersection(grid, pos, direction):
    dist = 0
    x, y = pos
    dx, dy = direction
    neighbors = set(get_neighbors(grid, pos)) - {(x - dx, y - dy)}
    while len(neighbors) == 1:
        neighbor = next(n for n in neighbors if n != (x - dx, y - dy))
        dx, dy = (neighbor[0] - x, neighbor[1] - y)
        x, y = neighbor
        dist += 1
        neighbors = set(get_neighbors(grid, (x, y))) - {(x - dx, y - dy)}
    return (x, y), dist + 1


def get_neighbors(grid: list[list[str]], pos: tuple[int, int]) -> list[tuple[int, int]]:
    for dx, dy in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
        x, y = (pos[0] + dx, pos[1] + dy)
        if not 0 <= x < len(grid) or not 0 <= y < len(grid[0]) or grid[x][y] == "#":
            continue
        yield (x, y)


def test_thomren():
    """
    Run `python -m pytest ./day-23/part-2/thomren.py` to test the submission.
    """
    assert (
        ThomrenSubmission().run(
            """
#.#####################
#.......#########...###
#######.#########.#.###
###.....#.>.>.###.#.###
###v#####.#v#.###.#.###
###.>...#.#.#.....#...#
###v###.#.#.#########.#
###...#.#.#.......#...#
#####.#.#.#######.#.###
#.....#.#.#.......#...#
#.#####.#.#.#########v#
#.#...#...#...###...>.#
#.#.#v#######v###.###v#
#...#.>.#...>.>.#.###.#
#####v#.#.###v#.#.###.#
#.....#...#...#.#.#...#
#.#########.###.#.#.###
#...###...#...#...#.###
###.###.#.###v#####v###
#...#...#.#.>.>.#.>.###
#.###.###.#.###.#.#v###
#.....###...###...#...#
#####################.#
""".strip()
        )
        == 154
    )

File: day-23/part-2/thomren.rs
use std::collections::HashMap;
use std::env::args;
use std::time::Instant;

fn main() {
    let now = Instant::now();
    let output = run(&args().nth(1).expect("Please provide an input"));
    let elapsed = now.elapsed();
    println!("_duration:{}", elapsed.as_secs_f64() * 1000.);
    println!("{}", output);
}

type Coord = (usize, usize);

fn run(input: &str) -> isize {
    let grid: Vec<&[u8]> = input.lines().map(|l| l.as_bytes()).collect();
    let start = (0, 1);
    let end = (grid.len() - 1, grid[0].len() - 2);

    // build condensed graph
    let mut stack = vec![start];
    let mut visited = vec![vec![false; grid[0].len()]; grid.len()];
    let mut graph: HashMap<Coord, Vec<(Coord, usize)>> = HashMap::new();
    while let Some(pos) = stack.pop() {
        if pos == end || visited[pos.0][pos.1] {
            continue;
        }
        visited[pos.0][pos.1] = true;

        for neighbor in get_neighbors(&grid, pos) {
            let (n, nd) = find_next_intersection(
                &grid,
                neighbor,
                (
                    neighbor.0 as isize - pos.0 as isize,
                    neighbor.1 as isize - pos.1 as isize,
                ),
            );
            (*graph.entry(pos).or_default()).push((n, nd));
            stack.push(n);
        }
    }

    // find longest path
    let mut stack = vec![(start, 0)];
    let mut visited = vec![vec![false; grid[0].len()]; grid.len()];
    let mut max = 0;
    while let Some((pos, dist)) = stack.pop() {
        // println!("{:?} {}", pos, dist);
        if pos == end {
            max = max.max(dist);
            continue;
        }
        if dist == -1 {
            visited[pos.0][pos.1] = false;
            continue;
        }
        if visited[pos.0][pos.1] {
            continue;
        }
        visited[pos.0][pos.1] = true;

        // will backtrack once all the paths from pos have been explored
        stack.push((pos, -1));

        for (neighbor, d) in graph[&pos].iter() {
            if !visited[neighbor.0][neighbor.1] {
                stack.push((*neighbor, dist + *d as isize));
            }
        }
    }
    max
}

fn find_next_intersection(
    grid: &Vec<&[u8]>,
    pos: (usize, usize),
    direction: (isize, isize),
) -> ((usize, usize), usize) {
    let mut dist = 0;
    let (mut x, mut y) = pos;
    let (mut dx, mut dy) = direction;
    let mut neighbors = get_neighbors(grid, pos)
        .into_iter()
        .filter(|&n| n != (x.wrapping_add_signed(-dx), y.wrapping_add_signed(-dy)))
        .collect::<Vec<(usize, usize)>>();
    while neighbors.len() == 1 {
        let neighbor = neighbors[0];
        dx = neighbor.0 as isize - x as isize;
        dy = neighbor.1 as isize - y as isize;
        (x, y) = neighbor;
        dist += 1;
        neighbors = get_neighbors(grid, (x, y))
            .into_iter()
            .filter(|&n| n != (x.wrapping_add_signed(-dx), y.wrapping_add_signed(-dy)))
            .collect::<Vec<(usize, usize)>>();
    }
    ((x, y), dist + 1)
}

fn get_neighbors(grid: &Vec<&[u8]>, pos: (usize, usize)) -> Vec<(usize, usize)> {
    let mut neighbors = vec![];
    if pos.0 > 0 && grid[pos.0 - 1][pos.1] != b'#' {
        neighbors.push((pos.0 - 1, pos.1));
    }
    if pos.0 < grid.len() - 1 && grid[pos.0 + 1][pos.1] != b'#' {
        neighbors.push((pos.0 + 1, pos.1));
    }
    if pos.1 > 0 && grid[pos.0][pos.1 - 1] != b'#' {
        neighbors.push((pos.0, pos.1 - 1));
    }
    if pos.1 < grid[0].len() - 1 && grid[pos.0][pos.1 + 1] != b'#' {
        neighbors.push((pos.0, pos.1 + 1));
    }
    neighbors
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn run_test() {
        assert_eq!(
            run("#.#####################
#.......#########...###
#######.#########.#.###
###.....#.>.>.###.#.###
###v#####.#v#.###.#.###
###.>...#.#.#.....#...#
###v###.#.#.#########.#
###...#.#.#.......#...#
#####.#.#.#######.#.###
#.....#.#.#.......#...#
#.#####.#.#.#########v#
#.#...#...#...###...>.#
#.#.#v#######v###.###v#
#...#.>.#...>.>.#.###.#
#####v#.#.###v#.#.###.#
#.....#...#...#.#.#...#
#.#########.###.#.#.###
#...###...#...#...#.###
###.###.#.###v#####v###
#...#...#.#.>.>.#.>.###
#.###.###.#.###.#.#v###
#.....###...###...#...#
#####################.#"),
            154
        )
    }
}
